packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -1551,43 +1576,75 @@ export class SnapController extends BaseController<
-          const [error, version] = resolveVersionRange(rawVersion);
-          if (error) {
-            result[snapId] = {
-              error: ethErrors.rpc.invalidParams(
-                `The "version" field must be a valid SemVer version range if specified. Received: "${version}".`,
-              ),
-            };
-            return;
-          }
-          const permissionName = getSnapPermissionName(snapId);
+    const snapIds = Object.keys(requestedSnaps);
 
-          if (
-            this.messagingSystem.call(
-              'PermissionController:hasPermission',
-              origin,
-              permissionName,
-            )
-          ) {
-            // Attempt to install and run the snap, storing any errors that
-            // occur during the process.
-            result[snapId] = {
-              ...(await this.processRequestedSnap(origin, snapId, version)),
-            };
+    // Existing snaps may need to be updated
+    const pendingUpdates = snapIds.filter((snapId) => this.has(snapId));
+
+    // Non-existing snaps will need to be installed
+    const pendingInstalls = snapIds.filter(
+      (snapId) => !pendingUpdates.includes(snapId),
+    );
+
+    try {
+      for (const entry of Object.entries(requestedSnaps)) {
+        const snapId = entry[0];
+        const { version: rawVersion } = entry[1];
+
+        const [error, version] = resolveVersionRange(rawVersion);
+
+        if (error) {
+          throw ethErrors.rpc.invalidParams(
+            `The "version" field must be a valid SemVer version range if specified. Received: "${rawVersion}".`,
+          );
+        }
+
+        const permissionName = getSnapPermissionName(snapId);
+
+        if (
+          !this.messagingSystem.call(
+            'PermissionController:hasPermission',
+            origin,
+            permissionName,
+          )
+        ) {
+          throw ethErrors.provider.unauthorized(
+            `Not authorized to install snap "${snapId}". Request the permission for the snap before attempting to install it.`,
+          );
+        }
+
+        const isUpdate = pendingUpdates.includes(snapId);
+
+        if (isUpdate && this.#isValidUpdate(snapId, version)) {
+          let rollbackSnapshot = this.#getRollbackSnapshot(snapId);
+          if (rollbackSnapshot === undefined) {
+            const prevSourceCode = this.#getRuntimeExpect(snapId).sourceCode;
+            rollbackSnapshot = this.#createRollbackSnapshot(snapId);
+            rollbackSnapshot.sourceCode = prevSourceCode;
+            rollbackSnapshot.newVersion = version;
           } else {
-            // only allow the installation of permitted snaps
-            result[snapId] = {
-              error: ethErrors.provider.unauthorized(
-                `Not authorized to install snap "${snapId}". Request the permission for the snap before attempting to install it.`,
-              ),
-            };
+            console.error('This snap is already being updated.');
```

Comment:
Should this throw an error instead?

Reply:
Yeah, I was thinking about this as well. I'll change this to a throw.